netty/src/main/java/io/grpc/netty/ClientTransportLifecycleManager.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import io.grpc.Status;
+import io.grpc.internal.ManagedClientTransport;
+
+/** Maintainer of transport lifecycle status. */
+final class ClientTransportLifecycleManager {
+  private final ManagedClientTransport.Listener listener;
+  private boolean transportReady;
+  private boolean transportShutdown;
+  /** null iff !transportShutdown. */
+  private Status shutdownStatus;
+  /** null iff !transportShutdown. */
+  private Throwable shutdownThrowable;
+  private boolean transportTerminated;
+
+  public ClientTransportLifecycleManager(ManagedClientTransport.Listener listener) {
+    this.listener = listener;
+  }
+
+  public void notifyReady() {
+    if (transportReady) {
+      return;
+    }
+    transportReady = true;
+    listener.transportReady();
+  }
+
+  public void notifyShutdown(Status s) {
+    if (transportShutdown) {
+      return;
+    }
+    transportReady = true; // Inhibit ready callback
```

Comment:
Rather than setting `transportReady = true`, why not just add conditions to `notifyReady` to return if `transportShutdown`?